flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/ZeroCopyNettyMessageDecoder.java

```diff
@@ -0,0 +1,280 @@
+    private DecodeStep decodeStep;
+
+    /** How many bytes have been decoded in current step. */
+    private int decodedBytesOfCurrentStep;
+
+    /** The intermediate state when decoding the current message. */
+    private final MessageDecodeIntermediateState intermediateState;
+
+    ZeroCopyNettyMessageDecoder(NettyMessageParser messageParser) {
+        this.messageParser = messageParser;
+        this.intermediateState = new MessageDecodeIntermediateState();
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        super.channelActive(ctx);
+
+        frameHeaderBuffer = ctx.alloc().directBuffer(NettyMessage.FRAME_HEADER_LENGTH);
+        messageHeaderBuffer = ctx.alloc().directBuffer(INITIAL_MESSAGE_HEADER_BUFFER_LENGTH);
+
+        decodeStep = DecodeStep.DECODING_FRAME_HEADER;
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        super.channelInactive(ctx);
+
+        if (intermediateState.messageHeaderParseResult != null) {
+            Buffer buffer = intermediateState.messageHeaderParseResult.getTargetDataBuffer();
+
+            if (buffer != null) {
+                buffer.recycleBuffer();
+            }
+        }
+
+        clearState();
+
+        frameHeaderBuffer.release();
+        messageHeaderBuffer.release();
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        if (!(msg instanceof ByteBuf)) {
+            ctx.fireChannelRead(msg);
+            return;
+        }
+
+        ByteBuf data = (ByteBuf) msg;
+
+        try {
+            while (data.readableBytes() > 0) {
+                if (decodeStep == DecodeStep.DECODING_FRAME_HEADER) {
+                    ByteBuf toDecode = cumulateBufferIfNeeded(frameHeaderBuffer, data, FRAME_HEADER_LENGTH);
+
+                    if (toDecode != null) {
+                        decodeFrameHeader(toDecode);
+
+                        decodedBytesOfCurrentStep = 0;
+                        decodeStep = DecodeStep.DECODING_MESSAGE_HEADER;
+                    }
+                }
+
+                if (decodeStep == DecodeStep.DECODING_MESSAGE_HEADER) {
+                    ByteBuf toDecoder = cumulateBufferIfNeeded(messageHeaderBuffer, data, intermediateState.messageHeaderLength);
+
+                    if (toDecoder != null) {
+                        intermediateState.messageHeaderParseResult = messageParser.parseMessageHeader(intermediateState.msgId, toDecoder);
+
+                        if (intermediateState.messageHeaderParseResult.getDataBufferAction() ==
+                                NettyMessageParser.DataBufferAction.NO_DATA_BUFFER) {
+                            ctx.fireChannelRead(intermediateState.messageHeaderParseResult.getParsedMessage());
+                            clearState();
+                        } else {
+                            decodedBytesOfCurrentStep = 0;
```

Comment:
ditto: This is reductant as commented in https://github.com/apache/flink/pull/7368/files#r376207885

Reply:
Have removed this line